### ISSUE TITLE
How to execute validation in parallel during the training phase in low-level Tensorflow API?

### ISSUE THREAD

**Issue by SUZhaoyu** [uber/horovod#599]
I play with Horovod for a while and it performs quite well for multi-server training. However, I have a question: how to enable validation right after each training epoch in low-level Tensorflow? I went through the docs and only found examples for high-level TF api such as estimator with validation, and the example in [tensorflow_wrod2vec.py](https://github.com/uber/horovod/blob/3c58a4da2a491cf503e591f340c9d05a2814fabe/examples/tensorflow_word2vec.py#L238) only executes the validation on one single worker.

I tried to hack the original example [tensorflow_mnist.py](https://github.com/uber/horovod/blob/master/examples/tensorflow_mnist.py) into something like below (pseudo code):

```
with with tf.train.MonitoredTrainingSession(some_configs) as mon_sess:
    while not mon_sess.should_stop():
        training_accuracy = train_one_epoch(mon_sess, training_data)
        validation_accuracy = valid_one_epoch(mon_sess, validation_data)
```

But is there anyway to collect the accuracy yielding from different workers together? And is there any build-in horovod api for validation in parallel? Thank you!

**Comment by tgaddair** [comment]
Hey @SUZhaoyu, validation will already be run in parallel the way you've set it up in your code.  If you want to aggregate the metrics together, you can either `allreduce` them (sum or average), or `[allgather](https://github.com/uber/horovod/blob/master/horovod/tensorflow/mpi_ops.py#L107)` (collect them into a list).

**Comment by SUZhaoyu** [comment]
Thank you for the reply :) @tgaddair, it seems that the `hvd.allgather()` or `hvd.allreduce()` requires the `tf.tensor` as its input. However, in my case the `validation_accuracy ` is actually a numpy object  yielded from `validation_accuracy = mon_sess.run(input_data)`. So where should I put the `mon_sess.run()`? Still place it within the `valid_one_epoch ` function or within the `if hvd.rank() == 0:`?

**Comment by tgaddair** [comment]
There are a few ways to convert a scalar into a tensor.  One way is to use [tf.convert_to_tensor](https://www.tensorflow.org/api_docs/python/tf/convert_to_tensor).  Another option for which we have an example is to wrap it in a `tf.Variable`.  This is what we do for the [MetricAverageCallback](https://github.com/uber/horovod/blob/master/horovod/keras/callbacks_impl.py#L33) class which does exactly what you're trying to do, but for Keras.

**Comment by SUZhaoyu** [comment]
Thanks a lot dude. I tried to use `tf.convert_to_tensor` but failed with the error: _'the graph has been finalized and cannot be modified.'_(Something similar to this, I cannot recall precisely). Later I tried to add the metric collection to the TF graph definition like this (again, pseudo code):

```
def tf_graph(graph_parameters):
    loss = ...
    loss_collection = hvd.allreduce(loss)
    return loss_collection
```

then execute the optimization based on the returned `loss_collection` and it works finally.

**Comment by SUZhaoyu** [comment]
Another issue is that, how can I plot the training and validation metrics in the tensorboard? One idea comes to my mind is to save the training and validation metrics separately (e.g., two folders), then activate tensorboard at the upper directory like [this](https://stackoverflow.com/questions/48951136/plot-multiple-graphs-in-one-plot-using-tensorboard). However, in horovod, the `summary_dir` has been prefixed via the [tf.train.MonitoredTrainingSession](https://www.tensorflow.org/api_docs/python/tf/train/MonitoredTrainingSession), which means that both training and validation metrics will share the same directory, making is impossible to plot figures like [this](http://cs231n.github.io/assets/nn3/accuracies.jpeg). Is it feasible to assign different `summary_dir` for training and validation?

**Comment by tgaddair** [comment]
Sure, you don't need to specify the summary dir or even use MonitoredTrainingSession if you'd prefer not to.  If you still need help with this, feel free to post a code snippet showing what you're trying to accomplish and I'll happily take a look.

**Comment by SUZhaoyu** [comment]
Sorry for the delayed reply. I tried to add different summary writers for different phases, and the problem has been solved. Thanks for your kind help, really appreciate! @tgaddair
